compendium/DeclarativeServices/src/manager/states/CCActiveState.cpp

```diff
@@ -34,51 +35,61 @@ std::shared_ptr<ComponentInstance> CCActiveState::Activate(
+        // throw inside LatchScopeGuard's dtor.
+        try {
+          latch.CountDown();
+        } catch (...) {
+          logger->Log(cppmicroservices::logservice::SeverityLevel::LOG_ERROR,
+                      "latch.CountDown() threw an exception during "
+                      "LatchScopeGuard cleanup.",
+                      std::current_exception());
+        }
+      });
+      std::lock_guard<std::mutex> lock(oneAtATimeMutex);
+      // Make sure the state didn't change while we were waiting
+      auto currentState = mgr.GetState();
+      if (currentState->GetValue() !=
+          service::component::runtime::dto::ComponentState::ACTIVE) {
+        logger->Log(cppmicroservices::logservice::SeverityLevel::LOG_WARNING,
+                    "Activate failed. Component no longer in Active State.");
+        return nullptr;
+      }
+      // no state change, already in active state. create and return a ComponentInstance object
+      // This could throw; a scope guard is put in place to call latch.CountDown().
+      instance = mgr.CreateAndActivateComponentInstance(clientBundle);
+
+      // Just in case the configuration properties changed between Registration and
+      // Construction of the component, update the properties in the service registration object.
+      // An example of when this could happen is when immediate=false and configuration-policy
+      // = optional. The component could be registered before all the configuration objects are
+      // available but it won't be constructed until someone gets the service. In between those
+      // two activities the configuration objects could change and the service registration properties
+      // would be out of date.
+      if (instance) {
+        mgr.SetRegistrationProperties();
+      }
+    }
+    if (!instance) {
+      logger->Log(cppmicroservices::logservice::SeverityLevel::LOG_ERROR,
+                  "Component configuration activation failed");
+    }
+    return instance;
   }
-  return instance;
+  // do not allow any new component instances to be created if Deactivate was called
+  logger->Log(cppmicroservices::logservice::SeverityLevel::LOG_DEBUG,
+              "Component configuration activation failed because component is "
+              "not in active state");
+  return nullptr;
 }
-
-void CCActiveState::Deactivate(ComponentConfigurationImpl& mgr)
+  void CCActiveState::Deactivate(ComponentConfigurationImpl& mgr)
 {
-  std::lock_guard<std::mutex> lock(oneAtATimeMutex);
-
-  // Make sure the state didn't change while we were waiting
-  if (mgr.GetState()->GetValue() !=
-      service::component::runtime::dto::ComponentState::ACTIVE) {
-    auto logger = mgr.GetLogger();
-    logger->Log(cppmicroservices::logservice::SeverityLevel::LOG_ERROR,
-                "Deactivate failed. Component no longer in Active State.");
-    return;
-  }
-  DoDeactivateWork(mgr);
+   DoDeactivateWork(mgr);
```

Comment:
It doesn't seem like there's a need for a `DoDeactivateWork()` function. `Deactivate` can be called instead.

Reply:
DoDeactivateWork is also called by the Modified method

Reply:
Modified can call Deactivate. I don't see a need for DoDeactiveWork when it just calls Deactivate now.